File: wallet.py
from datetime import datetime

class Wallet:
    def __init__(self):
        self.balance = 0
        self.transactions = []
        self.savings_goal = 0

    def display_balance(self):
        print(f"Your current balance is: {self.balance}")

    def add_funds(self):
        amount = int(input("Enter the amount to add: "))
        self.balance += amount
        self.transactions.append((datetime.now(), f"Added {amount} Ar to wallet"))
        print(f"Added {amount} Ar to wallet")

    def subtract_funds(self):
        amount = int(input("Enter the amount to subtract: "))
        if amount <= self.balance:
            self.balance -= amount
            self.transactions.append((datetime.now(), f"Subtracted {amount} Ar from wallet"))
            print(f"Subtract {amount} Ar from wallet")
        else:
            print("Insufficient funds. Transaction canceled.")

    def set_savings_goal(self):
        self.savings_goal = int(input("Enter your savings goal amount: "))
        print(f"Savings goal set to {self.savings_goal} Ar.")

    def track_savings_progress(self):
        print(f"Current savings progress: {self.balance} / {self.savings_goal} Ar")

    def get_all_transactions(self):
        for date, desc in self.transactions:
            print(f"{date}: {desc}")


if __name__ == '__main__':
    wallet = Wallet()

    print("Welcome to the wallet program")
    while True:
        print(
            '''
            1. Display Balance
            2. Add Funds
            3. Subtract Funds
            4. Set Savings Goal
            5. Track Savings Progress
            6. View All Transactions
            7. Quit
            '''
        )
        choice = int(input("Choose an option (1-7): "))

        if choice == 1:
            wallet.display_balance()
        elif choice == 2:
            wallet.add_funds()
        elif choice == 3:
            wallet.subtract_funds()
        elif choice == 4:
            wallet.set_savings_goal()
        elif choice == 5:
            wallet.track_savings_progress()
        elif choice == 6:
            wallet.get_all_transactions()
        elif choice == 7:
            print("Exiting wallet program. Goodbye!")
            break
        else:
            print("Invalid choice. Please choose a number between 1 and 7.")

File: wallet.go
package main

import (
	"fmt"
	"time"
)

type Wallet struct {
	balance         int
	transactions    []Transaction
	savingsGoal     int
}

type Transaction struct {
	date  time.Time
	desc  string
}

func (w *Wallet) displayBalance() {
	fmt.Printf("Your current balance is: %d Ar\n", w.balance)
}

func (w *Wallet) addFunds() {
	var amount int
	fmt.Print("Enter the amount to add: ")
	fmt.Scan(&amount)

	w.balance += amount
	w.transactions = append(w.transactions, Transaction{time.Now(), fmt.Sprintf("Added %d Ar to wallet", amount)})

	fmt.Printf("Added %d Ar to wallet\n", amount)
}

func (w *Wallet) subtractFunds() {
	var amount int
	fmt.Print("Enter the amount to subtract: ")
	fmt.Scan(&amount)

	if amount <= w.balance {
		w.balance -= amount
		w.transactions = append(w.transactions, Transaction{time.Now(), fmt.Sprintf("Subtracted %d Ar from wallet", amount)})
		fmt.Printf("Subtract %d Ar from wallet\n", amount)
	} else {
		fmt.Println("Insufficient funds. Transaction canceled.")
	}
}

func (w *Wallet) setSavingsGoal() {
	fmt.Print("Enter your savings goal amount: ")
	fmt.Scan(&w.savingsGoal)
	fmt.Printf("Savings goal set to %d Ar.\n", w.savingsGoal)
}

func (w *Wallet) trackSavingsProgress() {
	fmt.Printf("Current savings progress: %d / %d Ar\n", w.balance, w.savingsGoal)
}

func (w *Wallet) getAllTransactions() {
	for _, t := range w.transactions {
		fmt.Printf("%s: %s\n", t.date.Format(time.RFC3339), t.desc)
	}
}

func main() {
	wallet := Wallet{}

	fmt.Println("Welcome to the wallet program")
	for {
		fmt.Println(`
		1. Display Balance
		2. Add Funds
		3. Subtract Funds
		4. Set Savings Goal
		5. Track Savings Progress
		6. View All Transactions
		7. Quit
		`)

		var choice int
		fmt.Print("Choose an option (1-7): ")
		fmt.Scan(&choice)

		switch choice {
		case 1:
			wallet.displayBalance()
		case 2:
			wallet.addFunds()
		case 3:
			wallet.subtractFunds()
		case 4:
			wallet.setSavingsGoal()
		case 5:
			wallet.trackSavingsProgress()
		case 6:
			wallet.getAllTransactions()
		case 7:
			fmt.Println("Exiting wallet program. Goodbye!")
			return
		default:
			fmt.Println("Invalid choice. Please choose a number between 1 and 7.")
		}
	}
}
